library/choose_disk.py

```diff
@@ -0,0 +1,651 @@
+                disk_type = is_ceph_disk("/dev/" + partition)
+                if disk_type:
+                    if ceph_disk:
+                        ceph_disk += " + "
+                    # Saving the ceph type (journal and/or data)
+                    ceph_disk += disk_type
+
+            if not ceph_disk:
+                logger.info(' Ignoring %10s : Device has exisiting partitions', physical_disk)
+                continue
+
+        selected_devices[physical_disk] = physical_disks[physical_disk]
+        selected_devices[physical_disk]['bdev'] = '/dev/' + physical_disk
+
+        if ceph_disk:
+            selected_devices[physical_disk]['ceph_prepared'] = ceph_disk
+            logger.info(' Adding   %10s : Ceph disk detected (%s)', physical_disk, ceph_disk)
+        else:
+            logger.info(' Adding   %10s : %s', physical_disk, selected_devices[physical_disk]['bdev'])
+
+    return selected_devices
+
+
+def get_block_devices_persistent_name(physical_disks):
+    '''
+    Replace the short name (sda) by the persistent naming 'by-id'
+    '''
+    directory = "/dev/disk/by-id/"
+
+    logger.info('Finding persistent disks name')
+
+    # If the directory doesn't exist, reports the list as-is
+    if not os.path.isdir(directory):
+        logger.info(' Cannot open %s', directory)
+        return physical_disks
+
+    final_disks = {}
+    matching_devices = {}
+
+    # Searching in the /dev/disk/by-id/ which are the disks we look for
+    for f in os.listdir(directory):
+        device_name = os.readlink(directory + f).split("/")[-1]
+        if device_name in physical_disks:
+            if device_name not in matching_devices:
+                matching_devices[device_name] = [f]
+            else:
+                matching_devices[device_name].append(f)
+
+    # Saving matching disks and add the path of the block device
+    # Output is like : Renaming        vdb to                        virtio-e0b94c7d-75ca-4175-a
+    for physical_disk in sorted(physical_disks):
+        if physical_disk in matching_devices:
+            current_index = sorted(matching_devices[physical_disk])[0]
+            final_disks[current_index] = physical_disks[physical_disk]
+            final_disks[current_index]["bdev"] = "%s%s" % (directory, current_index)
+            logger.info(' Renaming %10s to %50s', physical_disk, current_index)
+        else:
+            current_index = physical_disk
+            final_disks[current_index] = physical_disks[physical_disk]
+
+    return final_disks
+
+
+def fake_device(legacy_devices, ceph_type):
+    '''
+    In case of legacy block device names, let's create an internal faked
+    entry with a 'bdev' entry filled with the actual path. This will be used to
+    make a match later on.
+    '''
+    devices = {}
+    count = 0
+    for device in legacy_devices:
+        devices["%s_%d" % (ceph_type, count)] = {"bdev": os.path.dirname(device)+"/"+os.path.basename(device)}
```

Comment:
paths that are getting joined like this should use the `os.path.join`, a more robust implementation of what this line is trying to do here. From the docs:

    Join two or more pathname components, inserting '/' as needed.
    If any component is an absolute path, all previous path components
    will be discarded.  An empty last part will result in a path that
    ends with a separator.